smk/gene_abundance.smk

```diff
@@ -567,35 +618,47 @@ rule gene_catalog_mapping_profiling:
         bwaindex_prefix=${{bwaindex_prefix%.0123}}
         (time (bwa-mem2 mem -a -t {threads} -v 3 ${{bwaindex_prefix}} $input_files | \
                 msamtools filter -S -b -l {params.length} -p {wildcards.identity} -z 80 --besthit - > aligned.bam) >& {output.bwa_log}
-        total_reads="$(grep Processed {output.bwa_log} | perl -ne 'm/Processed (\\d+) reads/; $sum+=$1; END{{printf "%d\\n", $sum/2;}}')"
-        #echo $total_reads
-        common_args="--label {wildcards.omics}.{params.sample_alias} --total=$total_reads --mincount={params.mapped_reads_threshold} --pandas"
-        msamtools profile aligned.bam $common_args -o profile.TPM.txt.gz --multi prop --unit tpm
-        msamtools profile aligned.bam $common_args -o profile.abund.all.txt.gz --multi all --unit abund --nolen
-
-        rsync -a aligned.bam {output.filtered}
-        rsync -a profile.TPM.txt.gz {output.profile_tpm}
-        rsync -a profile.abund.all.txt.gz {output.map_profile}
+            total_reads="$(grep Processed {output.bwa_log} | perl -ne 'm/Processed (\\d+) reads/; $sum+=$1; END{{printf "%d\\n", $sum/2;}}')"
+            #echo $total_reads
+
+            # Run multiple msamtools modes in parallel using GNU parallel.
+            # We cap it by 'threads' limit to parallel.
+            common_args="--label {wildcards.omics}.{params.sample_alias} --total=$total_reads --mincount={params.mapped_reads_threshold} --pandas"
+            parallel --jobs {threads} <<__EOM__
+msamtools profile aligned.bam $common_args -o profile.TPM.txt.gz --multi prop --unit tpm
+msamtools profile aligned.bam $common_args -o profile.abund.all.txt.gz --multi all --unit abund --nolen
+__EOM__
+
+            rsync -a aligned.bam {output.filtered}
+            rsync -a profile.TPM.txt.gz {output.profile_tpm}
+            rsync -a profile.abund.all.txt.gz {output.map_profile}
         ) >& {log}
         """
 
 ###############################################################################################
 # Computation of read counts to genes belonging to MAGs or publicly available genomes
 ###############################################################################################
 
+# Ideally, bedtools multicov is run in the same rule as bwa-mapping to avoid reading BAM file over NFS.
+# However, if the BAM file exists already, just running bedtools on the BAM files on NFS will be faster.
+# I have currently disabled this competition between the rules for making 'bed.gz' by renaming output file.
+# Could be turned back on in the future.
+
```

Comment:
Could this rule be removed in the future, if the other rule has no problems?

Reply:
Yes, we could. I didn't remove it to catch the edge cases when we want to just rerun multicov with updated BED without rerunning alignment. May be it is not so common, so we could get rid of it at some point.